rfcs/0081-dynmical-config-for-stream-job.md

```diff
@@ -0,0 +1,81 @@
+authors:
+  - "st1page"
+start_date: "2023/12/07"
+---
+
+# Dynamic Config For Stream Job
+
+Thanks all the people took part in the discussion of https://github.com/risingwavelabs/risingwave/issues/11929 and implantation of `stream_rate_limit`. The RFC is hight inspired by them.
+
+## Background
+Currently, we have lots of config, switch and parameter for the streaming execution. They often used for performance optimization for different workload. But they distributed different places currently.
+- hardcode as constant value, can not alter, such as 
+  - `STATE_CLEANING_PERIOD_EPOCH` https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/src/stream/src/common/table/state_table.rs#L65
+  - `TOPN_CACHE_HIGH_CAPACITY_FACTOR` https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/src/stream/src/executor/top_n/top_n_cache.rs#L30
+- in specified stream plan node's proto, such as 
+  - `OverWindowCachePolicy` https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/proto/stream_plan.proto#L676-L690, 
+  - `rate_limit` https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/proto/stream_plan.proto#L164
+    - And it can be altered with a special barrier command https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/proto/stream_plan.proto#L85
+- stored in the `StreamGraph` as a `StreamEnvironment`, such as 
+  - `timezone`. https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/proto/stream_plan.proto#L829-L832, https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/proto/meta.proto#L100
+    - It is expected not to alter, which will be explained below
+- stores as the global config such as 
+  - `max_dirty_groups_heap_size` https://github.com/risingwavelabs/risingwave/blob/a8aa905ef26ef8c00f55a7c5b8ce76e6a0f7b72d/src/common/src/config.rs#L801
+
+And in future we might introduce more parameters for streaming, such as `log_store_capacity`.
+
+In other database, those parameter is declared by session variable or session config. It is because the compute task is bounded in those batch system and it is not necessary to alter the config for a running job. But for a long-running streaming job, alter the config for a created and processing job is needed.
+
+This RFC will introduce a dynamic config framework to make user
+- can declare those configuration when create streaming job
+- can alter those configuration for a long-running streaming job. 
+### Non-Goals
+
+The RFC will not concern those config can change the streaming graph and its topology, such as
+- The configs that can change the plan such as `RW_TWO_PHASE_AGG`
+- The configs that can change the actor's number such as `STREAMING_PARALLELISM` (maybe we can use the same syntax to alter it later, but it will be only a semantic sugar of scale)
```

Comment:
```suggestion
- The configs that can change the actor's number such as `STREAMING_PARALLELISM` (maybe we can use the same syntax to alter it later, but it will be only a syntactic sugar of scale)
```